include/rapidjson/istreamwrapper.h

```diff
@@ -43,62 +43,82 @@ RAPIDJSON_NAMESPACE_BEGIN
+        \param buffer user-supplied buffer.
+        \param bufferSize size of buffer in bytes. Must >=4 bytes.
+    */
+    BasicIStreamWrapper(StreamType& stream, Ch* buffer, size_t bufferSize)
+        : stream_(stream), buffer_(buffer), bufferSize_(bufferSize),
+        bufferLast_(0), current_(buffer_), readCount_(0), count_(0), eof_(false)
+    {
+        RAPIDJSON_ASSERT(bufferSize >= 4);
+        Read();
     }
 
-    Ch Take() { 
-        typename StreamType::int_type c = stream_.get();
-        if (RAPIDJSON_LIKELY(c != StreamType::traits_type::eof())) {
-            count_++;
-            return static_cast<Ch>(c);
-        }
-        else
-            return '\0';
+    /*!
+        \param stream Stream to read from. This uses unbuffered access, which can be extremely slow.
+    */
+    BasicIStreamWrapper(StreamType& stream)
+        : stream_(stream), buffer_(fallbackBuffer), bufferSize_(sizeof(fallbackBuffer)),
+        bufferLast_(0), current_(buffer_), readCount_(0), count_(0), eof_(false)
+    {
+        Read();
     }
 
-    // tellg() may return -1 when failed. So we count by ourself.
-    size_t Tell() const { return count_; }
+    Ch Peek() const { return *current_; }
+    Ch Take() { Ch c = *current_; Read(); return c; }
+    size_t Tell() const { return count_ + static_cast<size_t>(current_ - buffer_); }
 
-    Ch* PutBegin() { RAPIDJSON_ASSERT(false); return 0; }
+    // Not implemented
     void Put(Ch) { RAPIDJSON_ASSERT(false); }
-    void Flush() { RAPIDJSON_ASSERT(false); }
+    void Flush() { RAPIDJSON_ASSERT(false); } 
+    Ch* PutBegin() { RAPIDJSON_ASSERT(false); return 0; }
     size_t PutEnd(Ch*) { RAPIDJSON_ASSERT(false); return 0; }
 
     // For encoding detection only.
     const Ch* Peek4() const {
-        RAPIDJSON_ASSERT(sizeof(Ch) == 1); // Only usable for byte stream.
-        int i;
-        bool hasError = false;
-        for (i = 0; i < 4; ++i) {
-            typename StreamType::int_type c = stream_.get();
-            if (c == StreamType::traits_type::eof()) {
-                hasError = true;
-                stream_.clear();
-                break;
-            }
-            peekBuffer_[i] = static_cast<Ch>(c);
-        }
-        for (--i; i >= 0; --i)
-            stream_.putback(peekBuffer_[i]);
-        return !hasError ? peekBuffer_ : 0;
+        return (current_ + 4 <= bufferLast_) ? current_ : 0;
     }
 
 private:
     BasicIStreamWrapper(const BasicIStreamWrapper&);
     BasicIStreamWrapper& operator=(const BasicIStreamWrapper&);
 
+    void Read() {
+        if (current_ < bufferLast_)
+            ++current_;
+        else if (!eof_) {
+            count_ += readCount_;
+            stream_.read(buffer_, (std::streamsize)bufferSize_);
+            readCount_ = (size_t)stream_.gcount();
+            bufferLast_ = buffer_ + readCount_ - 1;
+            current_ = buffer_;
+
+            if (readCount_ < bufferSize_) {
+                buffer_[readCount_] = '\0';
```

Comment:
All `StreamType`s (if any) do *not* garantee reading the asked number of bytes, it's not necessarily EOF here I think; maybe test `stream_.eof()` or `readCount_ <= 0` instead?

Otherwise looks good/valuable to me (FWIW)..